pkg/sharding/handler.go

```diff
@@ -0,0 +1,202 @@
+			return
+		}
+		// update vector clock, resource version in data, send
+		raw, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			klog.Warningf("error reading sharded request body: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		if err := resp.Body.Close(); err != nil {
+			klog.Warningf("error closing sharded request body: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		obj, _, err := unstructured.UnstructuredJSONScheme.Decode(raw, nil, nil)
+		if err != nil {
+			klog.Warningf("could not parse sharded request body as unstructured: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		switch t := obj.(type) {
+		case *unstructured.UnstructuredList:
+			if err := state.UpdateWith(shard, t); err != nil {
+				klog.Warningf("could not update sharded state: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			updatedResourceVersion, err := state.ToResourceVersions().Encode()
+			if err != nil {
+				klog.Warningf("could not format new resource version: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			t.SetResourceVersion(updatedResourceVersion)
+			updatedContinueToken, err := state.Encode()
+			if err != nil {
+				klog.Warningf("could not format new continue token: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			t.SetContinue(updatedContinueToken)
+			for _, item := range t.Items {
+				item.SetClusterName(fmt.Sprintf("%s/clusters/%s", shard, item.GetClusterName()))
+			}
+			body := bytes.Buffer{}
+			if err := unstructured.UnstructuredJSONScheme.Encode(t, &body); err != nil {
+				klog.Warningf("could not encode sharded response body: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			resp.Header.Set("Content-Length", strconv.Itoa(body.Len()))
+			copyInto(w, resp.Header, resp.StatusCode, ioutil.NopCloser(&body))
+			return
+		default:
+			klog.Warningf("could not cast returned object as a list: %#v", obj)
+		}
+	}
+}
+
+func delegateTo(cfg *rest.Config, continueToken string, req *http.Request) (*http.Response, error) {
+	// until we have a central authority for authn data, we're sitting behind the auth stack for the
+	// proximal kcp the client hit and using impersonation to make it look like everyone's in agreement
+	userInfo, ok := request.UserFrom(req.Context())
+	if !ok {
+		return nil, errors.New("no user info found in request context")
+	}
+
+	cfg = rest.CopyConfig(cfg)
+	cfg.Impersonate = rest.ImpersonationConfig{
+		UserName: userInfo.GetName(),
+		Groups:   userInfo.GetGroups(),
+		Extra:    userInfo.GetExtra(),
+	}
+	transport, err := rest.TransportFor(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("could not get transport for config: %w", err)
+	}
+
+	var httpClient *http.Client
+	if transport != http.DefaultTransport {
```

Comment:
I think we just always want to do this, both to enforce the timeout even in this rare case, and so that `httpClient.Do` doesn't nil-panic on line 242.